# APS106 - Fundamentals of Computer Programming

## Week 11 | Lecture 2 (11.2) - More Files, File Methods, CSV Module, Creating Your Own Modules

### Lecture Structure

1. [Writing a Dictionary to File](#section1)
2. [More File Reading Method](#section2)
3. [Reading Files and `\n`](#section3)
4. [The `with` Statement](#section4)
5. [Breakout Session 1](#section5)
6. [CSV Files](#section6)
7. [Breakout Session 2](#section7)
8. [Modules](#section8)
9. [Breakout Session 3](#section9)

<a id='section1'></a>

## 1. Writing a Dictionary to File

In the previous lecture, we learned how to open a file, write to it, and close it. Now, let's apply those skills to a more practical example. Suppose we have a dictionary of student names and their grades, and we want to save this data to a file so it can be accessed later. We can loop through the dictionary and write each key-value pair as a line in the file, separating the name and grade with a comma. This is a common pattern for storing structured data in text files.

In [ ]:
students = {'Kendrick': 'A+', 'Dre': 'C-', 'Snoop': 'B'} 

# create a file
myfile = open("grades.txt", "w")

# store dictionary items to the file
for student in students:
    myfile.write(student + ',' + students[student] + '\n')

# close the file
myfile.close()

<a id='section2'></a>

## 2. More File Reading Methods

So far, we have used `file.read()` to read the entire contents of a file into a single string. While this works well for small files, it is often more useful to process a file one line at a time. Python gives us two convenient ways to do this: iterating directly over the file object with a `for` loop, and using the `readlines()` method to get a list of lines. Both approaches let us work with individual lines, but they differ slightly in how they give us access to the data.

### `for line in file`

The simplest way to read a file line-by-line is to use a `for` loop directly on the file object. Each iteration of the loop assigns the next line of the file to the loop variable. Note that each line **includes the newline character `\n`** at the end, so you will often need to use `strip()` to remove it before processing.

In [2]:
# declaring filename to change easily if you want to try with different files
filename = 'grades.txt'

In [5]:
# Approach: for line in file
# When to use it: When you want to process the file line-by-line
# Example code

myfile = open(filename, 'r')
contents = ''
for line in myfile: # each time through the loop line contains one line of the file
    contents += line
    #print(line)  
    # why is there a gap between rows? because there is an \n at line end, and print() adds another \n
myfile.close()

print(contents)
# by the end of this contents contains the entire contents of the file

Kendrick,A+
Dre,C-
Snoop,B



### `readlines()`

The `readlines()` method reads the entire file and returns a **list of strings**, where each string is one line from the file. Just like with `for line in file`, each string in the list **includes the `\n` character** at the end. This is why you will typically use `strip()` when processing the lines. This approach is useful when you want to access lines by index or need to know how many lines are in the file before processing them.

In [7]:
# Approach: readlines
# When to use it: When you want to process the file line-by-line with an index
# Example code
myfile = open(filename, 'r')
lines = myfile.readlines() # lines is a list of strings. Each entry in lines is a line of the file
myfile.close()

print(lines)

['Kendrick,A+\n', 'Dre,C-\n', 'Snoop,B\n']


Now that we know how to read a file line-by-line, let's put it all together. We originally wrote a dictionary to `grades.txt` with each line formatted as `Name,Grade`. Let's read the file back in and reconstruct the dictionary. We can use `strip()` to clean each line and `split(',')` to separate each line into its name and grade components.

The `strip()` method removes all leading and trailing whitespace from a string, including spaces, `\n`, `\t`, and `\r`. This is important here because each line read from a file includes a newline character `\n` at the end. You can also pass a string argument to strip specific characters, for example `"##hello##".strip("#")` returns `"hello"`.

In [ ]:
students = {}
myfile = open("grades.txt", "r")

# read each line of the file
for line in myfile:
    # remove the newline character and split on the comma
    parts = line.strip().split(',')   # --> [name, grade]
    name = parts[0]
    grade = parts[1]
    students[name] = grade

myfile.close()

print(students)

{'Kendrick': 'A+', 'Dre': 'C-', 'Snoop': 'B'}


<a id='section3'></a>

## 3. Reading Files and `\n`

When Python reads text from a file, it usually keeps the newline character `\n` at the end of each line. This is important because the newline is invisible when you print normally, but it still affects comparisons, formatting, and processing. To make hidden characters visible, we can use `repr(...)`, which shows special characters such as `\n` inside quotes. `repr(...)` will not be covered on the exam.

In this section, we will use the file `names.txt`, which contains:

```text
Alice
Bob
Charlie
Diana
```

This file ends with a newline after each name, so all of the reading methods below will preserve `\n`.

### `read()`

`read()` returns the entire file as one string. Since the newline characters are hidden when printed normally, we will use `repr(...)` to show the raw string and make each `\n` visible.

In [9]:
myfile = open('names.txt', 'r')
contents = myfile.read()
myfile.close()

print(repr(contents))

'Alice\nBob\nCharlie\nDiana\n'


### `readline()`

`readline()` returns one line each time you call it, including the trailing `\n`. Using `repr(...)` lets us see that newline character directly. Don't worry about `repr(...)` for the exam, its just being use for demonstration purposes.

In [10]:
myfile = open('names.txt', 'r')

for i in range(4):
    line = myfile.readline()
    print(repr(line))

myfile.close()

'Alice\n'
'Bob\n'
'Charlie\n'
'Diana\n'


### `for line in file`

Looping directly over a file also gives you one line at a time, and each line still ends with `\n`. Again, `repr(...)` makes the hidden newline visible.

In [11]:
myfile = open('names.txt', 'r')

for line in myfile:
    print(repr(line))

myfile.close()

'Alice\n'
'Bob\n'
'Charlie\n'
'Diana\n'


### `readlines()`

`readlines()` returns a list of strings. Each string in the list is one line from the file, and each one still includes its trailing `\n`.

In [12]:
myfile = open('names.txt', 'r')
lines = myfile.readlines()
myfile.close()

print(lines)

['Alice\n', 'Bob\n', 'Charlie\n', 'Diana\n']


All four methods preserve the newline characters from the file. A common and simple way to remove the trailing newline is `strip()`.

The `strip()` method removes whitespace from the beginning and end of a string. Whitespace includes spaces, tabs, and newline characters like `\n`. It does **not** remove characters from the middle of the string.

When we read a line from a file, the main whitespace we usually want to remove is the newline at the end. That is why `strip()` is so useful when working with text files.

Here is a small example:

In [13]:
line = 'Alice\n'

print(repr(line))
print(repr(line.strip()))

'Alice\n'
'Alice'


And here is another example.

In [14]:
line = '    Alice\nSeb and Chris\n   '

print(repr(line))
print(repr(line.strip()))

'    Alice\nSeb and Chris\n   '
'Alice\nSeb and Chris'


<a id='section4'></a>

## 4. The `with` Statement

Notice that whenever we open a file, we need to be careful to close it again. Python provides a nice way to open and then automatically close a file using a `with` block.

```python
with open(filename, mode) as variable:
    body
```

The file is opened at the beginning and **automatically closed** at the end of the body.

Let's try using `with` to read the contents of `grades.txt`. Notice that we do not need to call `file.close()` - Python handles that for us automatically when the `with` block ends.

In [15]:
with open('grades.txt', 'r') as file:
    content = file.read()
    print(content)

Kendrick,A+
Dre,C-
Snoop,B



In [16]:
with open('grades.txt', 'r') as grades_file:
    for line in grades_file:
        print(line, end="")

Kendrick,A+
Dre,C-
Snoop,B


<a id='section5'></a>

## 5. Breakout Session 1

You have been given a file called `playlist.txt` that contains a playlist of classic songs from the 60s and 70s. Each line of the file is formatted as `Artist,Song Title,Duration` where the duration is in seconds. Here are the first few lines of the file:

```
The Beatles,Hey Jude,431
Led Zeppelin,Stairway to Heaven,482
Pink Floyd,Comfortably Numb,382
The Rolling Stones,Paint It Black,222
Jimi Hendrix,Purple Haze,170
...
```

Write a function called `longest_song(filename)` that reads the playlist file using a `with` statement and returns the name of the longest song.

In [29]:
def longest_song(filename: str) -> str:
    """
    (str) -> str
    Read a playlist file and return the name of the longest song.
    """
    longest_name = None
    longest_duration = 0
    
    with open(filename, 'r') as file:
        for line in file:
            #print(line)
            line_ls = line.split(',')
            name = line_ls[1]
            duration = int(line_ls[2])
            
            if duration > longest_duration:
                longest_duration = duration
                longest_name = name
    
    return str(longest_name)

#### Test Case
Answer: `Stairway to Heaven`

In [30]:
print(longest_song('playlist.txt'))

Stairway to Heaven


<a id='section6'></a>

## 6. CSV Files

The CSV format (comma-separated values) is very commonly used to represent the data in a spreadsheet. 

For example, a spreadsheet such as:

Name|Test1|Test2|Final
----|-----|-----|-----
Kendrick|100|50|29
Dre|76|32|33
Snoop|25|75|95

is represented as a file like this:

```
Name,Test1,Test2,Final
Kendrick,100,50,29
Dre,76,32,33
Snoop,25,75,95
```

We can, of course, access this file using the techniques above.

Let's start by creating a CSV file using basic file writing techniques. We can write the header row and data rows as comma-separated strings, just like we did with `grades.txt`. Later, we will see how Python's `csv` module makes this much easier.

In [ ]:
with open('grades.csv','w') as csv_file:
    csv_file.write('Name,Test1,Test2,Final\nKendrick,100,50,29\nDre,76,32,33\nSnoop,25,75,95')

In [32]:
with open('grades.csv', 'r') as file:
    for line in file:
        print(line, end='')

Name,Test1,Test2,Final
Kendrick,100,50,29
Dre,76,32,33
Snoop,25,75,95

Notice that you have the information about each row and also the commas. If you are going to process this data, you are going to need to **parse** it. That means, for example, to discard the commas (as they just separate the data and are not otherwise meaningful) and to extract the integers from the string.

One of the great things about Python is the existence of many modules that give us the ability to easily do many things, like reading and writing CSV files. 

Reading of CSV files can be done using the CSV reader. You can construct a reader object using `csv.reader()` which takes the file object as input. The reader object can be used to iterate through the contents of the CSV file, similarly to how a file object was used to iterate through the contents in a text file.

The difference between the two is that the file method `read()` returns the entire contents of the file as one long string, whereas the CSV `reader()` returns an object which can be iterated through. The reader object holds each row as a list of strings and can be iterated through row by row. 

Example: Read each row of a CSV file

In [33]:
import csv

with open('grades.csv', 'r') as file:
    print(file)
    grades_reader = csv.reader(file) # Create csv.reader object with an open file
    print(grades_reader)             # What does grades_reader look like? just some object, but it is iterable!
    for row in grades_reader:        # The csv.reader is an iterable!
        print(row)

<_io.TextIOWrapper name='grades.csv' mode='r' encoding='utf-8'>
['Name', 'Test1', 'Test2', 'Final']
['Kendrick', '100', '50', '29']
['Dre', '76', '32', '33']
['Snoop', '25', '75', '95']


### Converting `csv.reader` to a List

We can also convert the `csv.reader` object directly into a list of lists using `list()`. This gives us all the rows at once, which is useful when you want to access rows by index or skip the header row using slicing.

In [34]:
import csv

with open('grades.csv', 'r') as file:
    reader = csv.reader(file)
    data = list(reader)

print(data)
print()
print('Header:', data[0])
print('First student:', data[1])

[['Name', 'Test1', 'Test2', 'Final'], ['Kendrick', '100', '50', '29'], ['Dre', '76', '32', '33'], ['Snoop', '25', '75', '95']]

Header: ['Name', 'Test1', 'Test2', 'Final']
First student: ['Kendrick', '100', '50', '29']


### Writing CSV Files with `csv.writer`

Just as `csv.reader` helps us read CSV files, the `csv` module also provides a `csv.writer` object for writing data to CSV files. You create a writer object by passing an open file to `csv.writer()`, and then use the `writerow()` method to write one row at a time. Each row is provided as a list, and the writer handles inserting the commas for you.

In the example below, suppose there were a few marking errors on the final exam and we need to write an updated version of the grades to a new file.

In [ ]:
import csv

rows = [['Name', 'Test1', 'Test2', 'Final'],
        ['Kendrick', '100', '50', '69'],
        ['Dre', '76', '32', '53'],
        ['Snoop', '25', '75', '100']]

with open('grades_new.csv', 'w') as csvfile:
    
    print(csvfile)
    grades_writer = csv.writer(csvfile)
    print(grades_writer)

    for row in rows:
        grades_writer.writerow(row)

<a id='section7'></a>

## 7. Breakout Session 2

Using `csv.reader`, write a function called `add_averages(filename)` that reads `grades.csv` and returns a list of lists where each student's row has their average appended to the end.

Recall that `grades.csv` looks like this:

```
Name,Test1,Test2,Final
Kendrick,100,50,29
Dre,76,32,33
Snoop,25,75,95
```

In [ ]:
import csv

def add_averages(filename):
    """
    (str) -> list
    Read a CSV file of grades and return a list of lists
    with the average appended to each student's row.
    """
    results = []
    
    with open(filename, 'r') as file:
        reader = csv.reader(file)
        counter = 0
        for row in reader:
            if counter > 0:
                print(row)
                ...
            counter += 1
    
    return results

#### Test Case
Answer:
```python
[['Kendrick', '100', '50', '29', 59.67],
 ['Dre', '76', '32', '33', 47.0],
 ['Snoop', '25', '75', '95', 65.0]]
```

In [ ]:
print(add_averages('grades.csv'))

<a id='section8'></a>

## 8. Modules

### What is a Module?
A module in Python is a file containing Python definitions and statements. A module can define functions, classes, and variables, and it can also include executable code. You can use a module in your code by importing it. Python comes with many built-in modules, such as the `math`, `random`, `csv`, and `turtle` modules that we have learned about thus far in the course.

### Importing Modules
To use a module in your code, you need to import it. The syntax for importing a module is:

```python
import module_name
```
For example, let's import the `math` module.

In [35]:
import math

Once you have imported a module, you can use its functions and variables by prefixing them with the module name. For example, to use the `pi` constant from the `math` module, you would use:

In [36]:
print(math.pi)

3.141592653589793


Or, we can use functions to calculate the sine of an angle.

In [37]:
print(math.sin(math.radians(90)))

1.0


#### `from X import Y` Syntax

Sometimes, you might only want to import specific functions or variables from a module, rather than importing the entire module. In this case, you can use the `from X import Y` syntax:
```python
from module_name import variable_name
from module_name import function_name
from module_name import class_name
```
For example, to import only the `pi` constant from the `math` module, you would use:

In [38]:
from math import pi

print(pi)

3.141592653589793


This creates a variable named `pi` in the global space with the value `3.141592653589793` of type `float`.

In [39]:
type(pi)

float

You can also import multiple functions or variables from a module by separating them with commas, like this:

In [40]:
from math import pi, sqrt

print(pi)
print(sqrt(4))

3.141592653589793
2.0


In this case, there is now a function definition called `sqrt` in the global space that is callable.

You'll notice that when you import functions, classes, or variables in this way, you no longer need to access them using `math.`

#### `from X import *` Syntax
In Python, the `from X import *` syntax allows you to import all functions and variables from a module into your current namespace. This means that you can use the imported functions and variables without prefixing them with the module name (`math.`). For example, if you import all functions and variables from the `math` module, you could use `pi` directly, like this:

In [ ]:
from math import *

print(pi)

However, using `from X import *` is generally not recommended for several reasons.

1. It can make your code harder to read and understand. When you use the `from X import *` syntax, it's not always clear where a function or variable is coming from. This can make it difficult to trace the source of bugs and errors in your code.

2. It can lead to naming conflicts. If you import all functions and variables from two different modules that have the same name, you'll end up with naming conflicts. For example, if you import all functions and variables from both the `math` module and the `random` module, both define a `log` function with different behaviour. This can cause confusion and errors in your code.

3. It can make your code less efficient. When you use the `from X import *` syntax, Python has to load and compile all the functions and variables from the module, even if you're only going to use a few of them. This can slow down the loading time of your program and use up more memory than necessary.

In general, it's better to be explicit about which functions and variables you're importing from a module. This makes your code easier to read and understand, reduces the risk of naming conflicts, and can make your code more efficient. If you need to import a large number of functions and variables from a module, it's better to import them explicitly, like this:

In [41]:
from math import pi, sin, cos, tan

In [42]:
sin(pi)

1.2246467991473532e-16

This way, you can be sure which functions and variables you're using and avoid naming conflicts.

#### `import X as` Syntax
The default name given to modules in your code is the name of the module. For example, `math`, `random`, etc.

However, you can override the default module name using the `as` keyword. This is useful if the name of a module conflicts with a variable you have declared in your code. It is also useful if you want to reference another module that shares the same name.

Suppose we want to import the `turtle` module into our code but we already have a variable called `turtle`.

In [43]:
turtle = 'Stewart'

import turtle

print(turtle)

<module 'turtle' from 'c:\\Users\\xiang\\miniconda3\\Lib\\turtle.py'>


Now we have overwritten the variable `turtle` with the module `turtle`. To avoid this, we would use an alias when importing the `turtle` module. The syntax for importing a module with an alias is:

```python
import module_name as alias_name
```

In [44]:
turtle = 'Stewart'

import turtle as t

print(turtle)

Stewart


Now, when we want to use the `turtle` module, we can use the alias to access the functions, classes, and variables in the `turtle` module.

In [ ]:
new_turtle = t.Turtle()  #running this may cause your Jupyter kernel to crash...

### How to Create Your Own Module in Python

Creating your own module is useful when the variables, functions, or classes you're creating in a `.py` file or `.ipynb` file are generic and likely to be used across many notebooks, projects, assignments, etc. For example, it would be silly to continually create `sine` functions in every notebook. This is why someone created the `math` module.

A module in Python is simply a file with a `.py` extension that contains Python code. To create a module, you just need to write your code in a `.py` file and save it with an appropriate name. When you import a module in Python, the interpreter searches for the module in the following locations:

- The current directory
- The directories listed in the `PYTHONPATH` environment variable
- The installation-dependent default directory

To ensure Python is able to find our modules, we will be placing our `.py` files in the same folder as this `.ipynb` file. 

```
├── lectures
│   ├── week11
│   |   ├── lecture2
│   |   |   ├── notebooks
│   |   |   |   ├── week11_lecture2.ipynb
│   |   |   |   ├── geometry_calculations.py
│   |   |   ├── slides
```
If you're using JupyterHub, you can see that `geometry_calculations.py` is placed in the same folder as `week11_lecture2.ipynb`. If you're downloading these files from Quercus, please place them in the same folder.

Here is the code we have in `geometry_calculations.py`:

```python
import math


def rectangle_area(length, width):
    """
    (number, number) -> number
    Calculate the area of a rectangle.
    """
    return length * width


def circle_area(radius):
    """
    (number) -> number
    Calculate the area of a circle.
    """
    return math.pi * radius**2


def cylinder_volume(radius, height):
    """
    (number, number) -> number
    Calculate the volume of a cylinder.
    """
    return math.pi * radius**2 * height


def rectangular_prism_volume(length, width, height):
    """
    (number, number, number) -> number
    Calculate the volume of a rectangular prism (3D rectangle).
    """
    return length * width * height


def cylinder_surface_area(radius, height):
    """
    (number, number) -> number
    Calculate the surface area of a cylinder.
    """
    return 2 * math.pi * radius * (radius + height)


def rectangular_prism_surface_area(length, width, height):
    """
    (number, number, number) -> number
    Calculate the surface area of a rectangular prism (3D rectangle).
    """
    return 2 * (length*width + length*height + width*height)
```

We have six functions and we have also imported the `math` module to access the `pi` constant.

Now, you can use this module in your other Python programs by importing it. We can import `geometry_calculations` and use its functionality:

In [45]:
import geometry_calculations

radius = 5
area = geometry_calculations.circle_area(radius)

print('The area of a circle with radius ', radius, ' is ', area, '.', sep='')

The area of a circle with radius 5 is 78.53981633974483.


You can see all the functions in `geometry_calculations.py` by pressing TAB below.

In [46]:
geometry_calculations.math  #<--- place your cursor after the . and press TAB.

<module 'math' (built-in)>

Interestingly, because we imported the `math` module in `geometry_calculations.py`, we can access it as follows:

In [47]:
geometry_calculations.math.pi

3.141592653589793

<a id='section9'></a>

## 9. Breakout Session 3

Create a module called `basic_math` with the following four functions: `add`, `subtract`, `multiply`, and `divide`. The `divide` function should return `"Error. You cannot divide by zero."` if the user tries to divide by 0. Import your module.

In [54]:
import basic_math

#### Test Case 1
Answer: `3`

In [55]:
print(basic_math.add(1, 2))

3


#### Test Case 2
Answer: `4`

In [56]:
print(basic_math.subtract(5, 1))

4


#### Test Case 3
Answer: `6`

In [57]:
print(basic_math.multiply(2, 3))

6


#### Test Case 4
Answer: `0.5`

In [58]:
print(basic_math.divide(6, 12))

0.5


#### Test Case 5
Answer: `Error. You cannot divide by zero.`

In [59]:
print(basic_math.divide(6, 0))

Error. You cannot divide by zero.
